whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -0,0 +1,202 @@
+
+
+@Service
+public class RdapRelationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RdapRelationService.class);
+
+    private final Ipv4Tree ip4Tree;
+    private final Ipv6Tree ip6Tree;
+    private final Ipv4DomainTree ipv4DomainTree;
+    private final Ipv6DomainTree ipv6DomainTree;
+    private final RpslObjectDao rpslObjectDao;
+
+    public RdapRelationService(final Ipv4Tree ip4Tree, final Ipv6Tree ip6Tree,
+                               final Ipv4DomainTree ipv4DomainTree, final Ipv6DomainTree ipv6DomainTree,
+                               final RpslObjectDao rpslObjectDao) {
+        this.ip4Tree = ip4Tree;
+        this.ip6Tree = ip6Tree;
+        this.ipv4DomainTree = ipv4DomainTree;
+        this.ipv6DomainTree = ipv6DomainTree;
+        this.rpslObjectDao = rpslObjectDao;
+    }
+
+    public List<String> getDomainRelationPkeys(final String pkey, final RelationType relationType){
+        final Domain domain = Domain.parse(pkey);
+        final IpInterval reverseIp = domain.getReverseIp();
+        final List<IpEntry> ipEntries = getIpEntries(getIpDomainTree(reverseIp), relationType, reverseIp);
+
+        return ipEntries
+                .stream()
+                .map(ipEntry -> rpslObjectDao.getById(ipEntry.getObjectId()).getKey().toString())
+                .toList();
+    }
+
+    public List<String> getInetnumRelationPkeys(final String pkey, final RelationType relationType){
+        final IpInterval ip = IpInterval.parse(pkey);
+        final List<IpEntry> ipEntries = getIpEntries(getIpTree(ip), relationType, ip);
+        return ipEntries
+                .stream()
+                .map(ipEntry -> ipEntry.getKey().toString())
+                .toList();
+    }
+
+    private List<IpEntry> getIpEntries(final IpTree ipTree, final RelationType relationType,
+                                       final IpInterval reverseIp) {
+        return switch (relationType) {
+            case UP -> List.of(searchFirstLessSpecific(ipTree, reverseIp));
+            case TOP -> searchTopLevel(ipTree, reverseIp);
+            case DOWN -> ipTree.findFirstMoreSpecific(reverseIp);
+            case BOTTOM -> searchMostSpecificFillingOverlaps(ipTree, reverseIp);
+        };
+    }
+
+    private List<IpEntry> searchMostSpecificFillingOverlaps(final IpTree ipTree, final IpInterval reverseIp){
+        final List<IpEntry> mostSpecificValues = ipTree.findMostSpecific(reverseIp);
+        final List<? extends IpInterval<?>> ipResources = mostSpecificValues.stream().map(ip -> IpInterval.parse(ip.getKey().toString())).toList();
+        final Set<IpEntry> mostSpecificFillingOverlaps = Sets.newConcurrentHashSet();
+        ipResources.forEach(ipResource -> processChildren(ipTree, reverseIp, ipResource, mostSpecificFillingOverlaps));
+        return mostSpecificFillingOverlaps.stream().toList();
+    }
+
+    private static void processChildren(final IpTree ipTree, final IpInterval reverseIp,
```

Comment:
does this needs to be static?

Reply:
no needed

Reply:
I am filling mostSpecificFillingOverlaps list processing the leafs
It is not an easy name